mavros/src/plugins/setpoint_global_position.cpp

```diff
@@ -0,0 +1,192 @@
+	private plugin::SetPositionTargetLocalNEDMixin<SetpointGlobalPositionPlugin>{
+public:
+	SetpointGlobalPositionPlugin() : PluginBase(),
+		// NOTE not private handle, because we need to access gps topic, which is not under this node. Is it OK?
+		sp_nh("~")
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		/* subscriber for current gps state, mavros/global_position/global. */
+		gps_sub = sp_nh.subscribe("global_position/global", 10, &SetpointGlobalPositionPlugin::gps_cb, this);
+		/* Subscribe for current local pose. */
+		local_sub = sp_nh.subscribe("local_position/pose", 10, &SetpointGlobalPositionPlugin::local_cb, this) ;
+		/* Subscriber for target gps state */
+		setpoint_sub = sp_nh.subscribe("setpoint_global_position/global", 10, &SetpointGlobalPositionPlugin::setpoint_cb, this);
+
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return { /* Rx disabled */ };
+	}
+
+private:
+	friend class SetPositionTargetLocalNEDMixin;
+	ros::NodeHandle sp_nh;
+
+	ros::Subscriber setpoint_sub;
+	ros::Subscriber gps_sub;
+	ros::Subscriber local_sub;
+
+	/* resulting change in x/y in NED frame. Set by lla2ned() function. */
+	double dx, dy;
+
+	/* Stores current gps state. */
+	sensor_msgs::NavSatFix gps_msg;
+	/* Current local pose */
+	geometry_msgs::PoseStamped local_pose_msg;
+	/* old time gps time stamp in [ms], to check if new gps msg is received */
+	uint32_t old_gps_stamp = 0;
+
+	/* -*- mid-level helpers -*- */
+
+	/**
+	 * @brief Converts lat/long to 'change' in local NED frme
+	 * @param curr_lat. Current latitude in deg
+	 * @param curr_long. Current longitude in deg
+	 * @param target_lat. Target latitude in deg
+	 * @param target_long. Target longitude in deg
+	 */
+	 void lla2ned(double curr_lat, double curr_long, double target_lat, double target_long){
+		 /* Convert to radins */
+		double curr_lat_rad = curr_lat * M_DEG_TO_RAD;
+		double curr_lon_rad = curr_long * M_DEG_TO_RAD;
+
+		double curr_sin_lat = sin(curr_lat_rad);
+		double curr_cos_lat = cos(curr_lat_rad);
+
+		/* Convert target lat/long to radians */
+		double lat_rad = target_lat * M_DEG_TO_RAD;
+		double lon_rad = target_long * M_DEG_TO_RAD;
+
+		double sin_lat = sin(lat_rad);
+		double cos_lat = cos(lat_rad);
+
+		/* cos the difference in long */
+		double cos_d_lon = cos(lon_rad - curr_lon_rad);
+
+		double arg = curr_sin_lat * sin_lat + curr_cos_lat * cos_lat * cos_d_lon;
+		if (arg > 1.0)
+			arg = 1.0;
+		else if (arg < -1.0)
+			arg = -1.0;
+
+		double c = acos(arg);
+
+		double k=0.0;
+		if(abs(c) < DBL_EPSILON)
+			k=1.0;
+		else
+			k=c/sin(c);
+
+
+		dx = k * (curr_cos_lat * sin_lat - curr_sin_lat * cos_lat * cos_d_lon) * CONSTANTS_RADIUS_OF_EARTH;
+		dy = k * cos_lat * sin(lon_rad - curr_lon_rad) * CONSTANTS_RADIUS_OF_EARTH;
+	 }
+
+
+
+	/* -*- callbacks -*- */
+
+	void gps_cb(const sensor_msgs::NavSatFix::ConstPtr &msg){
+		gps_msg.latitude = msg->latitude;
+		gps_msg.longitude = msg->longitude;
+		gps_msg.altitude = msg->altitude;
```

Comment:
Also is your GPS giving the altitude on AMSL or is it the height over the WGS-84 ellipsoid? You need to add a parameter that considers the second as the NavSatFix considers the height over the ellipsoid. Agaim check `fake_gps` and `global_position` plugins to see how to convert from one altitude to another.